gno.land/pkg/sdk/vm/keeper.go

```diff
@@ -971,6 +995,128 @@ func (vm *VMKeeper) QueryDoc(ctx sdk.Context, pkgPath string) (*doc.JSONDocument
 	return d.WriteJSONDocumentation()
 }
 
+// QueryStorage returns storage and deposit for a realm.
+func (vm *VMKeeper) QueryStorage(ctx sdk.Context, pkgPath string) (string, error) {
+	store := vm.newGnoTransactionStore(ctx) // throwaway (never committed)
+	rlm := store.GetPackageRealm(pkgPath)
+	if rlm == nil {
+		err := ErrInvalidPkgPath(fmt.Sprintf(
+			"realm not found: %s", pkgPath))
+		return "", err
+	}
+	res := fmt.Sprintf("storage: %d, deposit: %d", rlm.Storage, rlm.Deposit)
+
+	return res, nil
+}
+
+// processStorageDeposit processes storage deposit adjustments for package realms based on
+// storage size changes tracked within the gnoStore.
+//
+// For each realm, it:
+// - Charges the caller a deposit proportional to newly used storage (positive size difference).
+// - Returns the deposit to the caller for released storage (negative size difference).
+//
+// An error is returned if there's insufficient deposit, a transfer error occurs,
+// or an invariant violation happens during deposit or storage adjustments.
+
+func (vm *VMKeeper) processStorageDeposit(ctx sdk.Context, caller crypto.Address, deposit std.Coins, gnostore gno.Store, params Params) error {
+	realmDiffs := gnostore.RealmDiffs()
+	depositAmt := deposit.AmountOf(ugnot.Denom)
+	if depositAmt == 0 {
+		depositAmt = std.MustParseCoin(params.DefaultDeposit).Amount
+	}
+	price := std.MustParseCoin(params.StoragePrice)
+
+	for rlmPath, diff := range realmDiffs {
+		if diff == 0 {
+			continue
+		}
+
+		rlm := gnostore.GetPackageRealm(rlmPath)
+
+		if diff > 0 {
+			// lock deposit for the additional storage used.
+			requiredDeposit := overflow.Mulp(diff, price.Amount)
+			if depositAmt < requiredDeposit {
+				return fmt.Errorf("not enough deposit to cover the storage usage: requires %d%s for %d bytes", requiredDeposit, ugnot.Denom, diff)
```

Comment:
We should find a way to return a single error with all the storage requirement issues and not stop at the first one. Else the UX will be terrible.

Reply:
Could you please clarify what you mean by "all the storage requirement issues and not stop at the first one"? How does this approach affect the user experience? If possible, could you provide some examples to illustrate the impact?

Reply:
like using multierr to aggregate these requirements for the whole “for loop”.